Implements a new way to deal with primary keys

HavokInspiration opened the pull request:
This PR aims to provide a way to deal with how Phinx automatically creates an ``id`` field.
Currently, the only way to do so is to define, for each ``table()`` calls, this array of options :

```php
$table = $this->table('statuses',
    [
        'id' => false,
        'primary_key' => ['id']
    ]);
```

With these changes, a new ``AbstractMigration`` class has been added.
It can accept a ``autoId`` property that, when set to ``false``, will prevent Phinx from automatically create the ``id`` column.
When set, you will have to manually create the ``id`` column.
A new ``autoIncrement`` option has been added so columns can be defined as such (it really is just a "sugar" option, it can currently be done using the ``identity`` option but I find it not self-explanatory).

This file is a good example of how to use this :
https://github.com/HavokInspiration/migrations/blob/custom-migrations/tests/test_app/config/TestsMigrations/20150826191400_create_letters_table.php

I also added a ``disable-autoid`` boolean option to the ``MigrationSnashot`` bake command. By default, baking a snapshot will generate a file similar to the one currently baked (except for the parent ``AbstractMigration`` class).
With this option, it will set the ``autoId`` property to false and use the new methods implemented to bake the snapshot.

**Note that when ``autoId`` is set to ``true`` in a migration class file, you need to manually create the PK column for all tables in the migration class file.**

Potential use cases : 
- when only UUID are used as PK
- if you want to customize the column(s) of the PK (for example, if you need it to be unsigned)
- if you want your PK column to not be named id

I'm completely open to suggestion, this does not really "improve" the potential "cumbersome-factor" of calling 

```php
$table = $this->table('statuses',
    [
        'id' => false,
        'primary_key' => ['id']
    ]);
```

for each table since every PK column now need to be defined manually if ``autoId`` is set to ``false`` (well except for UUID since you already have to do this) but the resulting migrations files, are, in my opinion, much more readable this way.

Could resolve #9 since the issue seems to be around that.

markstory commented on src/Table.php:
```diff
@@ -0,0 +1,67 @@
+        if ((!isset($this->options['id']) || $this->options['id'] === false) && !empty($this->primaryKey)) {
+            $this->options['primary_key'] = $this->primaryKey;
+        }
+        parent::create();
```
Does update need to be changed as well? Could people change the primary key and expect it to work the same as create?

HavokInspiration replied:
I'll have to check if this is needed but I think changing the primary key can already be done by removing the index, removing / renaming the column and adding the new primary key.

HavokInspiration replied:
Yup, you are right, there are cases where this is going to be needed.
I'll patch the PR.

HavokInspiration commented:
Just pushed a CS & SQLite build fix.
I'll take care of modifying the ``Table::update()`` method

HavokInspiration commented:
@markstory Actually, it is currently not possible to change a primary key on an ``update`` call. You can change the name of the column (with a ``renameColumn()`` call).
However, let's say you create a table without a PK in a migration. If you want to add a PK later on, you currently can not do it.

I'll dig to see if I can come up with something to fix this.

HavokInspiration commented:
I took a quick look and I think I can not easily do this on our side.
I might have to make a patch in phinx. I'll keep you posted.

HavokInspiration commented:
I don't know if this is something that should be done. SQLite does not support adding a primary key to an existing table.
Other db servers support it but that sound like something that would be to big of a difference with the others. On the other side, adding a PK after a table has been created sounds like bad practice to me.

lorenzo commented:
@HavokInspiration Yeah, it can be really tricky to manage

HavokInspiration commented:
I still need to take care of the baked migrations files that are not snapshots.
The current implementation of adding a primary key does not create a primary key.

HavokInspiration commented:
I pushed some commits.

One of them corrects the snapshot for tables having composite PK (the old writting was still used).

The others takes care of migrations creation using ``bin/cake bake migration create_users {{fields}}``.

Currently, if you bake using ``id:integer:primary_key``, it creates a syntax that will not work (at least with MySQL).
These commits fix this by using the new implementation.

Things to notice : 

- If a primary key is declared in the arguments, the migration files will set the ``autoId`` property to false and use the ``addPrimaryKey()`` method.
- If you define a PK as an integer (or without a type which will make it an integer by default), the column will be tagged as auto increment
- It works with composite PK : if you use ``id:primary_key name:string:primary_key`` it will create a composite PK, id being an int auto increment

lorenzo commented:
There a couple coding standards problems, apart from that this looks really good!

HavokInspiration commented:
Yup, I'll address them when #118 has bee taken cared of

HavokInspiration commented:
I fixed the CS errors, updated the tests from #118 and made the bake shell error out if there is a PK being added to an existing table, telling the user that adding a PK to an existing table is not supported.

HavokInspiration commented on src/Shell/Task/MigrationTask.php:
```diff
@@ -16,6 +16,7 @@
 use Cake\Core\Configure;
 use Cake\Event\Event;
 use Cake\Event\EventManager;
+use Cake\Utility\Hash;
```
This should be removed

lorenzo commented:
Great job!
